File: 0204-count-primes/0204-count-primes.py
"""
class Solution:
    def countPrimes(self, n: int) -> int:
        if n < 3:
            return 0
        if n == 3:
            return 1
        
        primes = 1
        for i in range(3, n):
            prime = True
            for j in range(2, i):
                if i % j == 0:
                    prime = False
                    break
            if prime:
                primes += 1
        
        return primes
"""

class Solution:
    def countPrimes(self, n: int) -> int:
        if n < 3:
            return 0
        
        table = [True] * n
        table[0], table[1] = False, False
        
        for i in range(2, n):
            if table[i]:
                cur = i * i
                while cur < n:
                    table[cur] = False
                    cur += i
        
        return sum(table)
                    

File: 0204-count-primes/0204-count-primes.go
const (
    prime = 0
    composite = 1
)

func countPrimes(n int) int {
    if n < 3 {
        return 0
    }
    
    primeTable := make([]int, n)
    primeTable[0], primeTable[1] = composite, composite
    for i := 2; i < n; i++ {
        if primeTable[i] == prime {
            multiple := i * i
            for multiple < n {
                primeTable[multiple] = composite
                multiple += i
            }
        }
    }
    
    primeCount := 0
    for _, v := range primeTable {
        if v == prime {
            primeCount ++
        }
    }
    return primeCount
}